### PR TITLE
Avoid unnecessary DDP synchronization when gradient_accumulation_steps > 1

### PR DESCRIPTION
**Pull request by noamwies**
# What does this PR do?

This PR avoid unnecessary ```DistributedDataParallel``` synchronization when gradient_accumulation_steps > 1 by using  ```DistributedDataParallel.no_sync```.

This lead to speedup when training with multiple gpu's for example the ```run_language_modeling.py``` complete wiki-2 epoch in 85 seconds instead of 111

```bash
python run_language_modeling.py --output_dir=runs --model_type=gpt2 --model_name_or_path=gpt2 --per_device_train_batch_size 6 --do_train --train_data_file=$TRAIN_FILE  --gradient_accumulation_steps=32 --fp16 --block_size 513 --overwrite_output_dir
```

## Before submitting
- [x]  Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests), 
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests? 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -101,6 +101,11 @@
     _use_native_amp = True
     from torch.cuda.amp import autocast
 
+if version.parse(torch.__version__) < version.parse("1.2"):
+    _use_ddp_no_sync = False
+else:
+    _use_ddp_no_sync = True
+
```
Comment (LysandreJik) [inline src/transformers/trainer.py:108]: We only support pytorch > 1.3.1 for the examples, so you can remove this if/else clause!

**Review by LysandreJik — APPROVED** [review]
Seems like textbook usage of the `no_sync()` method, LGTM!

File: src/transformers/trainer.py (hunk 1 above)
Comment (noamwies) [reply src/transformers/trainer.py]: Great, so I removed this check

Comment (sgugger) [reply src/transformers/trainer.py]: This is not an example though, and `Trainer` is in the main lib. So we should keep this check until we pin pytorch to a later version IMO.

**Review by sgugger — APPROVED** [review]
Thanks for adding this!

File: src/transformers/trainer.py (hunk 1 above)
Comment (LysandreJik) [reply src/transformers/trainer.py]: @sgugger is right.

Comment (noamwies) [reply src/transformers/trainer.py:108]: I reverted to the original implementation

Comment (LysandreJik) [reply src/transformers/trainer.py:108]: Thank you!

**Pull request merged**